daemon/containerd/image_list.go

```diff
@@ -46,6 +61,54 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 	return ret, nil
 }
 
+func (i *ImageService) setupFilters(ctx context.Context, opts types.ImageListOptions) ([]imageFilterFunc, error) {
```

Comment:
```suggestion
func (i *ImageService) setupFilters(ctx context.Context, opts types.ImageListOptions) (imageFilterFunc, error) {
```
```go
return func(image containerd.Image) bool {
        for _, filter := range filters {
                if !filter(image) {
                    return false
                }
        }
        return true
}, nil
```

There is only one filter. It may have more than one clause, but they're all logically parts of one filter. The caller of `setupFilters` shouldn't need to care how the filter is implemented; returning a slice of filter clauses just opens up opportunities for the calling code to evaluate the filter incorrectly (e.g. by ORing the clauses when they should be ANDed).

Reply:
Good suggestion; I was also looking at that.

- I updated the function as suggested
- Changed the signature to accept `filters.Args` instead of the whole `ImageListOptions`
- I moved the `i.setupFilters()` invocation to the start of the function, as parsing the filters should be less work than first fetching all images (only to return an error if there's an issue with constructing the filters)
- I added a second commit to validate the filters (as not all filters are implemented yet)
- Created some tracking issues for that; see  https://github.com/moby/moby/pull/43816#discussion_r926449121